include/dice/template-library/standard_layout_tuple.hpp

```diff
@@ -0,0 +1,139 @@
+
+#include <dice/template-library/type_list.hpp>
+#include <dice/template-library/type_traits.hpp>
+
+#include <cstddef>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace dice::template_library {
+
+	struct uniform_construct_t {};
+	inline constexpr uniform_construct_t uniform_construct;
+
+	struct individual_construct_t {};
+	inline constexpr individual_construct_t individual_construct;
+
+	/**
+	 * A tuple that fulfills std::is_standard_layout
+	 */
+	template<typename ...Ts>
+	struct standard_layout_tuple;
+
+	template<>
+	struct standard_layout_tuple<> {
+		constexpr standard_layout_tuple() noexcept = default;
+		explicit constexpr standard_layout_tuple(individual_construct_t) noexcept {}
+
+		template<typename ...Args>
+		explicit constexpr standard_layout_tuple(uniform_construct_t, [[maybe_unused]] Args &&...args) noexcept {}
+
+		static constexpr size_t size() noexcept {
+			return 0;
+		}
+
+		template<size_t offset, size_t count>
+		[[nodiscard]] standard_layout_tuple const &subtuple() const noexcept {
+			static_assert(offset == 0 && count == 0);
+			return *this;
+		}
+
+		constexpr auto operator<=>(standard_layout_tuple const &other) const noexcept = default;
+		constexpr bool operator==(standard_layout_tuple const &other) const noexcept = default;
+	};
+
+	template<typename T, typename ...Ts>
+	struct standard_layout_tuple<T, Ts...> {
+	private:
+		T first;
+		[[no_unique_address]] standard_layout_tuple<Ts...> rest;
+
+	public:
+		constexpr standard_layout_tuple() = default;
+
+		template<typename ...Args>
+		explicit constexpr standard_layout_tuple(Args &&...args) requires (sizeof...(Args) == 1 + sizeof...(Ts))
+			: standard_layout_tuple{individual_construct, std::forward<Args>(args)...}
+		{
+		}
+
+		template<typename Arg, typename ...Args>
+		explicit constexpr standard_layout_tuple(individual_construct_t, Arg &&first, Args &&...rest) requires (sizeof...(Ts) == sizeof...(Args))
+			: first{std::forward<Arg>(first)},
+			  rest{std::forward<Args>(rest)...}
+		{
+		}
+
+		template<typename ...Args>
+		explicit constexpr standard_layout_tuple(uniform_construct_t, Args &&...args)
+			: first{args...},
+			  rest{uniform_construct, std::forward<Args>(args)...}
+		{
+		}
+
+		[[nodiscard]] static constexpr size_t size() noexcept {
+			return 1 + sizeof...(Ts);
+		}
+
+		template<size_t ix, typename Self>
+		[[nodiscard]] constexpr decltype(auto) get(this Self &&self) noexcept {
+			static_assert(ix < size(), "Index for get must be in range");
+
+			if constexpr (ix == 0) {
+				return dice::template_library::forward_like<Self>(self.first);
+			} else {
+				return dice::template_library::forward_like<Self>(self.rest).template get<ix - 1>();
+			}
+		}
+
+		template<typename U, typename Self>
+		[[nodiscard]] constexpr decltype(auto) get(this Self &&self) noexcept {
+			using types = type_list::type_list<T, Ts...>;
+			static_assert(type_list::count_v<types, U> == 1,
```

Comment:
should this not be checked at construction?

Reply:
This is a general pattern, use `requires` if you want to remove the function from the overload set, use `static_assert` if there is no overload set and you just want to error on invalid usage.